linera-base/src/data_types.rs

```diff
@@ -948,6 +948,38 @@ pub enum BlobContent {
     ServiceBytecode(CompressedBytecode),
 }
 
+/// Compressed contract, service and bytecode_id.
+#[cfg(not(target_arch = "wasm32"))]
+#[derive(Clone)]
+pub struct CompressedContractService {
```

Comment:
A contract plus a service is an "application" --> `CompressedApplication`

Reply:
(I didn't try to see if we need this, though.)

Reply:
I think we need that struct because:
* The `{CompressedContractService,CompressedApplication}::new` is a significant piece of code with 3 arrays.
* Some of the runtime gain is about the compression, but there is also the gain coming from the use of the hash computation.
* The hashing is significant as we are hashing the bytecode so it takes some time.
* Having the two entries needlessly complicate the API of the `published_compressed_bytecode`.

What I can do is rename that struct to `HashedCompressedBytecode` which better reflects the goals of the struct.

Reply:
I'm not sure I understand. What 3 arrays? Also AFAIK we don't have a `CompressedApplication`.
How are we hashing any less by using this struct? 🤔 From my understanding you're moving compression and hashing further up the stack, but that has nothing to do with the struct.
The struct is just an abstraction that was created together with it, which I think it's unnecessary.

> Having the two entries needlessly complicate the API of the `published_compressed_bytecode`.

I actually think that the _struct_ needlessly complicates the API of the `published_compressed_bytecode` 😅 it's another layer of abstraction the code reader needs to go through and understand, and it doesn't has much benefit IMHO.

Maybe I'm missing something (please let me know if I am), but from my current understanding I believe I still disagree this is necessary.

Reply:
In the preceding code, the function calls `f(client).await` are done repeatedly.

Therefore, if this function is called several times, we also call the hash functions several times. Those calls are not needed, as the data has not changed. Therefore, by using the struct, we can avoid the rehash and thus avoid repeating a computation.

We have 3 data to pass and thus it makes sense to group them into a struct.

Reply:
I still don't get what's the difference from using the struct and just having `new` be instead a function that returns a tuple `(Blob, Blob, BytecodeId)` for example 🤔
I think this PR makes sense, I just don't think adding this extra struct abstraction is really necessary

Reply:
Won't the single clone operation be doing internally 3 clone operations in the same way? We just won't have to call `.clone()` explicitly 3 times, but that's still what will be happening internally AFAIU
This is not a huge deal, but I do think it makes the code harder to read as it adds an extra layer of abstraction. Also in all the rest of the code we don't really combine contract/service blobs in a single struct anywhere, we always use them separately.
So if improving readability is your point, I disagree. There also doesn't seem to be a performance gain.

Reply:
Doing the `.clone()` operation on a struct has the same cost as doing 3 `.clone()` operations. I do not disagree with that, I was thinking about the complexity of the code.

That being said, there is an argument that we have tried to separate `contract` and `service` separate so far. So, I will implement the function with returning `(Blob, Blob, BytecodeId)` as you asked.